File: frontend/src/domain/audio/value-objects/rms-db.ts
/**
 * RMS dB value object.
 */

export class RmsDb {
    private constructor(public readonly value: number) {
        if (value > 0) {
            throw new Error('RMS dB cannot be positive');
        }
        if (value < -120) {
            throw new Error('RMS dB cannot be below -120dB (essentially silence)');
        }
    }

    static create(value: number): RmsDb {
        return new RmsDb(value);
    }

    static fromLinear(linearValue: number): RmsDb {
        if (linearValue <= 0) {
            return new RmsDb(-120.0);
        }
        const dbValue = 20 * Math.log10(linearValue);
        return new RmsDb(dbValue);
    }

    isSilence(thresholdDb: number = -60.0): boolean {
        return this.value < thresholdDb;
    }

    isClipping(thresholdDb: number = -0.5): boolean {
        return this.value > thresholdDb;
    }

    equals(other: RmsDb): boolean {
        return Math.abs(this.value - other.value) < 0.01;
    }
}

File: backend/streamcraft/domain/audio/value_objects/rms_db.py
"""RMS dB value object."""

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class RmsDb:
    """RMS (Root Mean Square) in decibels."""

    value: float

    def __post_init__(self) -> None:
        """Validate RMS dB value."""
        if self.value > 0:
            raise ValueError("RMS dB cannot be positive")
        if self.value < -120:
            raise ValueError("RMS dB cannot be below -120dB (essentially silence)")

    @classmethod
    def from_linear(cls, linear_value: float) -> "RmsDb":
        """Create from linear scale value (0.0 to 1.0)."""
        import math
        if linear_value <= 0:
            return cls(value=-120.0)
        db_value = 20 * math.log10(linear_value)
        return cls(value=db_value)

    def is_silence(self, threshold_db: float = -60.0) -> bool:
        """Check if this represents silence."""
        return self.value < threshold_db

    def is_clipping(self, threshold_db: float = -0.5) -> bool:
        """Check if this represents clipping."""
        return self.value > threshold_db
